File: app-kotlin/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/pet/PETFragment.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.pet;

import android.os.Build;
import android.widget.PopupWindow;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.NetworkUtils;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.shared.GlobalPreferencesViewModel;
import com.google.firebase.analytics.FirebaseAnalytics;

public abstract class PETFragment extends Fragment {

    protected FirebaseAnalytics analytics;

    protected GlobalPreferencesViewModel globalPreferencesViewModel;

    @Nullable
    protected PopupWindow popupWindow;

    protected PETFragment() {
    }

    protected PETFragment(int layout) {
        super(layout);
    }

    protected void init() {
        setOnBackPressed();

        initViewModels();
        initFirebaseAnalytics();
    }

    protected abstract void initViewModels();

    protected void initGlobalPreferencesViewModel() {
        try {
            if (globalPreferencesViewModel == null) {
                globalPreferencesViewModel =
                        new ViewModelProvider(requireActivity()).get(GlobalPreferencesViewModel.class);

                globalPreferencesViewModel.init(requireContext());
            }
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    /**
     * saveStates method
     */
    protected void saveGlobalPreferencesViewModel() {
        try {
            if (globalPreferencesViewModel != null) {
                    globalPreferencesViewModel.saveToFile(requireContext());
            }
        } catch (IllegalStateException e){
            e.printStackTrace();
        }
    }

    /**
     * refreshFragment
     */
    protected void refreshFragment() {

        FragmentTransaction ft = getParentFragmentManager().beginTransaction();
        if (Build.VERSION.SDK_INT >= 26) {
            ft.setReorderingAllowed(false);
        }
        ft.detach(PETFragment.this).commitNow();
        ft = getParentFragmentManager().beginTransaction();
        ft.attach(PETFragment.this).commitNow();

    }

    protected void backPressedHandler() {
        if(closePopup()) {
            return;
        }

        try {
            Navigation.findNavController(requireView()).popBackStack();
    } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    protected boolean closePopup() {
        if(popupWindow != null && popupWindow.isShowing()) {
            popupWindow.dismiss();
            return true;
        }

        return false;
    }

    protected void setOnBackPressed() {
        try {
            requireActivity().getOnBackPressedDispatcher()
                    .addCallback(getViewLifecycleOwner(), new OnBackPressedCallback(true) {
                        @Override
                        public void handleOnBackPressed() {
                            backPressedHandler();
                        }
                    });
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }

    }

    protected void initFirebaseAnalytics() {
        try {
            this.analytics = ((PETActivity) requireActivity()).getFirebaseAnalytics();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    protected boolean checkInternetConnection() {
        if(globalPreferencesViewModel == null) { return false; }

        try {
            return (NetworkUtils.isNetworkAvailable(requireContext(),
                    globalPreferencesViewModel.getNetworkPreference()));
        } catch (IllegalStateException e) {
            e.printStackTrace();
            return false;
        }
    }

    protected void saveStates() {
        saveGlobalPreferencesViewModel();
    }
}

File: app-view/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/activities/pet/PETFragment.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.pet

import android.os.Build
import android.widget.PopupWindow
import androidx.activity.OnBackPressedCallback
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation.findNavController
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.PermissionsViewModel
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.GlobalPreferencesViewModel
import com.tritiumgaming.phasmophobiaevidencepicker.util.NetworkUtils.isNetworkAvailable
import com.google.firebase.analytics.FirebaseAnalytics
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.impl.AccountManagementService


abstract class PETFragment : Fragment, AccountManagementService {

    /*protected var globalPreferencesViewModel: GlobalPreferencesViewModel? = null
    protected var permissionsViewModel: PermissionsViewModel? = null*/

    protected val globalPreferencesViewModel: GlobalPreferencesViewModel by activityViewModels { GlobalPreferencesViewModel.Factory }
    protected val permissionsViewModel: PermissionsViewModel by activityViewModels()

    protected var analytics: FirebaseAnalytics? = null
    protected var popupWindow: PopupWindow? = null

    protected constructor()

    protected constructor(layout: Int) : super(layout)

    protected fun init() {
        setOnBackPressed()

        initViewModels()
        initFirebaseAnalytics()
    }

    protected fun initGlobalPreferencesViewModel() {
        globalPreferencesViewModel.init(requireContext())
    }

    private fun initFirebaseAnalytics() {
        try { this.analytics = (requireActivity() as PETActivity).firebaseAnalytics }
        catch (e: IllegalStateException) { e.printStackTrace() }
    }

    protected fun saveGlobalPreferencesViewModel() {
        try { globalPreferencesViewModel?.saveToFile(requireContext())
        } catch (e: IllegalStateException) { e.printStackTrace() }
    }

    //@SuppressLint("DetachAndAttachSameFragment")
    protected open fun refreshFragment() {
        var ft = parentFragmentManager.beginTransaction()
        if (Build.VERSION.SDK_INT >= 26) { ft.setReorderingAllowed(false) }
        ft.detach(this@PETFragment).commitNow()
        //ft = parentFragmentManager.beginTransaction()
        ft = parentFragmentManager.beginTransaction()
        if (Build.VERSION.SDK_INT >= 26) { ft.setReorderingAllowed(false) }
        ft.attach(this@PETFragment).commitNow()
    }

    protected fun closePopup(): Boolean {
        popupWindow?.let { popupWindow ->
            if (popupWindow.isShowing) {
                popupWindow.dismiss()
                return true
            }
        }

        return false
    }

    protected fun checkInternetConnection(): Boolean {
        return globalPreferencesViewModel?.let { globalPreferencesViewModel ->
            try {
                return (isNetworkAvailable(
                    requireContext(), globalPreferencesViewModel.networkPreference))
            } catch (e: IllegalStateException) {
                e.printStackTrace()
                return false
            }
        } ?:  return false
    }

    protected open fun saveStates() {
        saveGlobalPreferencesViewModel()
    }

    protected open fun backPressedHandler() {
        if (closePopup()) { return }

        try { findNavController(requireView()).popBackStack() }
        catch (e: IllegalStateException) { e.printStackTrace() }
    }

    private fun setOnBackPressed() {
        try {
            requireActivity().onBackPressedDispatcher
                .addCallback(viewLifecycleOwner, object : OnBackPressedCallback(true) {
                    override fun handleOnBackPressed() { backPressedHandler() }
                })
        } catch (e: IllegalStateException) { e.printStackTrace() }
    }

    protected abstract fun initViewModels()

}
